### ISSUE TITLE
Linux login manager & PAM integration for SSO

### ISSUE THREAD

**Issue by vorstrelok** [keepassxreboot/keepassxc#6055]
## Summary
Not quite sure where to post this, I hope you'll forgive me if it's a wrong place.
I've been using KeePassXC for a couple of weeks and I find typing my password 2 times on every login or screen unlock VERY inconvenient. I find https://github.com/keepassxreboot/keepassxc/issues/1404 and what is proposed by Q&A (use secret-tool to store password) not suitable, because:
1. It requires running additional service for storing password just to store your KeePassXC DB password
2. You can't use KeePassXC as you freedesktop secret service provider https://github.com/keepassxreboot/keepassxc/issues/1403

This issue has already been discussed here: https://github.com/keepassxreboot/keepassxc/issues/2797 and PAM integration will be needed for https://github.com/keepassxreboot/keepassxc/issues/3856 and https://github.com/keepassxreboot/keepassxc/issues/5991

Writing a PAM module would've been a proper solution, but it would require quite some time and learning.
So I spent some time (around 4 days probably) for possible workarounds and I created one!

## How it works
Upon login your user password is saved into your Linux user keyring (not to be confused with gnome or kde keyring!), then I have a service that monitors this keyring for a specifically named key and send this key via D-Bus call to running instance of KeePassXC. KeePassXC must be already running. End result is that is your KeePassXC will be automatically unlocked on login and screed unlock, while staying locked if your screen is locked. It is still a hack, but imo less hacky than using other secret service provider just for automatic unlock.

## How to do it
It needs keyutils and d-bus packages but you probably already have them. Oh and I wrote in fish shell script, but porting to other shell should be trivial.
Code is here: https://gist.github.com/vorstrelok/3b7a2b43c85e2b8b6cd3f81abe7a93fa
1. Put keepassxc_pam.fish and keepassxc_sendpass.fish somewhere (I chose my home folder for now)
2. Add this 2 lines to your /etc/pam.d/system-local-login or copy mine from the gist.
`auth      optional  pam_exec.so expose_authtok /home/vorstrel/keepassxc_pam.fish`
`session   optional  pam_exec.so /home/vorstrel/keepassxc_pam.fish`
Do not forget to change `/home/vorstrel` (my home folder name) to a name of your home folder! $HOME will not work, it's too early in session creation to have this variable and PAM config is not a shell scripting language.
Order of lines IS IMPORTANT, be sure to add each line after 'auth' and 'session' section like in my example!
Also, if you completely mess your /etc/pam.d/system-local-login you probably won't be able to login, so be sure to have recovery usb lying around just in case.  But you will probably be fine!
3. Add keepassxc_sendpass.fish to your prefered service autostart method. I used systemd user unit, example in the gist. Do not forget to change `/home/vorstrel` to your home folder
4. Done! 

## Future development
I will rewrite this in POSIX shell if there will be at least 3 users of my hack.
If this post will get 10 likes I will consider writing a proper PAM module if I will be able to recall how to code in C. No promises tho!
Hope my labour will be useful not just for me and help someone else too!

**Comment by phoerious** [comment]
This is pretty much out of scope. If anything, we could query the database password from the user's Gnome/KDE/macOS keyring, but then the whole security depends on the security of that and your login password is usually much less secure than your database password.

**Comment by droidmonkey** [comment]
I am absolutely against this integration. Quick unlock should solve most inconvenience imposed in a far more secure way. Agree that sharing the same security threshold as your computer account is poor practice.

**Comment by phoerious** [comment]
Also, Windows is still using NTHash, which basically MD4 of your UTF-16 password, so I would NEVER share a password with my password database on Windows.

**Comment by vorstrelok** [comment]
> Also, Windows still uses NTHash, which basically MD4 of your UTF-16 password, so I would NEVER share a password with my password database on Windows.

But I want this feature for linux?

> If anything, we could query the database password from the user's Gnome/KDE/macOS keyring

But if you use KeePassXC as your freedesktop secret service provider it won't work, because Gnome/KDE already provides freedesktop secret service! I mean, it creates circular dependency:

1. Keepassxc DB password is stored in 'secret service'
2. But you need to unlock Keepassxc to provide 'secret service'

I mean, I want to use KeePassXC as my keyring instead of Gnome/KDE etc. You already implemented freedesktop secret service to make it work. I use login password as my keepassxc db password. It is long and random and hard to guess etc. All that good security stuff. I just don't want to type it 2 times.
Gnome/KDE keyring (that are supposed to be your password manager) do the very same thing that I propose, use your user account password or some other authentication (fingerprint) to unlock keyring.

When you use `sudo` or `ssh` into your computer, you use your login account password and not some other password, don't you?

I am no security or PAM expert, but afaik if one of the apps that you use on X11 linux is compromised, your computer is screwed. Installing keylogger or changing your .bashrc so the next time you `sudo` password is sent somewhere is trivial.
To increase security you have to use Wayland based compositor & some form of LSM (AppArmor or SELinux)
Writing PAM module will allow securely transfer your credentials to KeePassXC so it can't be spied upon other programs running as your user (in my understanding)

> Quick unlock should solve most inconvenience imposed in a far more secure way.

Sorry if I missed something, but what do you mean by "Quick unlock"?

**Comment by droidmonkey** [comment]
I would think the whole reason you want keepassxc to be your secrets provider is that it's not coupled to your computer account like gnome keyring is. Similar to why you use keepassxc plus a browser plugin instead of storing passwords directly in the browser. The separation is part of the security model.

I would be open to allow unlock of keepassxc through dbus. Then you could use PAM on your own without need for integration.

**Comment by vorstrelok** [comment]
My current solution can't be considered a POC because it is a bunch of scripts to automatically and unsecurely send password to KeePassXC.
POC can be something like a PAM module that will receive login password, launch KeePassXC via D-Bus or some other way if need be, and attach user's login password in form of a linux key (not to be confused with gnome/kde keyring) to a running KeePassXC. This way eavesdropping from other applications (even running as your user UID) is impossible, if I understood linux keyring manpage correctly. You can read more about this facility here:
https://man.archlinux.org/man/keyrings.7
https://man.archlinux.org/man/keyctl.1

In this issue I wanted to ask if there is an interest in developing a proper integration and may be start developing it myself for inclusion in stock KeePassXC. You seem to be against it. Probably because you use 2 different password, one for login and another for keepass. I want one password, that is a promise of password manager, at least for me.
Even thou I disagree with you, I can accept your decision. You are the maintainer after all. :smiley: 

Thank you for developing KeePassXC, your program motivated me to finally move away from passwords stored in text files everywhere, I will continue to use it. :smile:

**Comment by droidmonkey** [comment]
I wouldn't be against someone doing this work and owning it. We don't want to maintain a PAM module specific to keepassxc. If there is a more generic PAM module that we can leverage that would be preferred. The communication plumbing should be similarly standard.

**Comment by phoerious** [comment]
KeePassXC isn't running when you are logging into your system via PAM, so we would need to store the hash in a kernel key ring and play it back or something (not so sure if that is a good idea), so I am not quite sure how this is supposed to work at the moment. You could of course authenticate through PAM again once it's running, but that wouldn't be any different from where we're at already.
Also keep in mind that Linux hashes passwords as SHA512. While that may be secure enough for most intents and purposes, it surely is a downgrade from Argon2.

**Comment by michaelk83** [comment]
> I would think the whole reason you want keepassxc to be your secrets provider is that it's not coupled to your computer account like gnome keyring is. ... The separation is part of the security model.

That really depends on the use case. In my case, I'm hoping to use KPXC as a drop-in replacement for KWallet, which hasn't been actively developed in years, suffers from a bunch of problems, and is effectively end-of-life ([ref1](https://mail.kde.org/pipermail/plasma-devel/2016-July/055641.html), [ref2](https://invent.kde.org/frameworks/kwallet/-/merge_requests/11#note_187369)). I also like the KPXC UI a lot more than KWallet, and I like that it's a Qt application, so doesn't add GNOME/Gtk or Mono dependencies.

Granted, this isn't entirely possible yet, since not all KDE apps have migrated to QtKeyChain yet (the ones that don't, still depend directly on KWallet), but I don't mind entering their passwords manually until they migrate. The general direction in KDE is to migrate to QtKeyChain (which supports `libsecret` as one of the backend options).

So my plan is:
1. Set KPXC as the system keyring, and disable KWallet.
2. Set up *separate* databases for the local passwords and more sensitive stuff.
3. Encourage KDE to continue their migration to QtKeyChain.

I don't mind the local passwords DB to have the same password as my user account (and `sudo`, as aptly noted by OP), since I've been doing that for years with KWallet. Given KWallet's current state, that would still be an upgrade. And given the same password (which can still be quite strong), it would be a nice convenience to *not* have to enter it twice at login.

I've seen auto-unlock-on-login requested multiple times here, so I and OP aren't the only ones who want this. The usually proposed solution of using `secret-tool` and `--pw-stdin` won't work if KPXC *is* the secret service provider, as OP has explained.

The other database would, of course, still have a separate password, and would *not* be auto-unlocked.

**Comment by michaelk83** [comment]
> KeePassXC isn't running when you are logging into your system via PAM, so we would need to store the hash in a kernel key ring and play it back or something

I'm not quite sure how PAM works, but assuming that:
1. The local passwords DB in KPXC has the same password as the user account.
2. That password has to be entered to log in.
3. It has to be entered again to resume a locked session.

Then I *would think* that it doesn't need to be stored *anywhere* (other than the mandatory `/etc/shadow`) for any longer than it takes to send it over DBus (or whichever) to unlock the passwords DB. Of course, if the DB is auto-locked after a few minutes, and then needed again, the password will have to be re-entered.

(Also worth noting that this should be disabled by default, and likely won't work anyway if the DB password is different from the user account. But it would still be nice to have that option.)

**Comment by developer91234** [comment]
> I don't mind the local passwords DB to have the same password as my user account (and sudo, as aptly noted by OP), since I've been doing that for years with KWallet. Given KWallet's current state, that would still be an upgrade. And given the same password (which can still be quite strong), it would be a nice convenience to not have to enter it twice at login.

Root passwords are going to be weaker than the Keepass key. They are typed in far more often, so they need to be easy to type. The whole point of using PAM is I think to avoid having to type in the long Keepass passkey every login. But if the root password and the keepass password are the same, then that's going to be even worse because now you're going to have to type in a very complex password to do anything that requires root/login.

This is what I do right now - I just have a helper database to unlock the main db, then use a script to open the main and close the helper.

PAM should unlock some intermediate layer which will in turn unlock the keepass database on the machine. The login pass and the keeXC database pass should be independent of each other.

**Comment by michaelk83** [comment]
If you're not sharing the DB online (e.g. backing it up to Google Drive), then having a helper DB with the user password, or having the main (but *offline-only*) DB with the user password, I think both cases have the same security. If the helper DB is compromised, so is the main one. And the point was, if your root password is compromised (which with `sudo` is usually the same as the user password), you have far bigger problems to worry about (*and* you have no way to protect the main DB in that case, anyway).

If you *are* sharing the main DB online, then yes, it makes sense to have a stronger passphrase on it, to protect it from outside your system. My approach for that is similar, but simpler: just have a separate DBs, and *don't* share the one with the user password.

**Comment by developer91234** [comment]
So you keep the db's in sync somehow?

**Comment by michaelk83** [comment]
In my particular case, I don't have anything in the local DB that I need to sync to the main one. But sure, if you have a lot of stuff that needs to be in both places, then the helper DB approach may be more convenient.

**Comment by danisztls** [comment]
I don't care for passwords or encryption on KeepassXC side as I already have an encryption layer on top of anything and my local system is hardened. If I have to protect my passwords from my local environment it's game over already. KeepasXC is great but does not need this forced attrition. Also it does not make sense because unless you have a Yubikey you are probably using a low entropy password or a keyfile residing on your userspace.

**Comment by phoerious** [comment]
> I don't care for passwords or encryption on KeepassXC side as I already have an encryption layer on top of anything and my local system is hardened.

If you don't need encryption, why do you use KeePassXC and not a notes manager or so? The point of KeePassXC is to encrypt your passwords. Everything else is added on. 

> Also it does not make sense because unless you have a Yubikey you are probably using a low entropy password or a keyfile residing on your userspace.

That's plain wrong. A simple passphrase can be very high-entropy and Argon2 prevents efficient bruteforcing. A key file can add additional protection for free, although a Yuibikey is of course the best option.

**Comment by danisztls** [comment]
> If you don't need encryption, why do you use KeePassXC and not a notes manager or so? Everything else is added on.

No notes manager that I know of has the same features for managing password as KeepasXC has. :smile: 

It's not that I don't want encryption. KeepasXC database is already under two encryption layers and having a third it's not an issue but just a nuisance as I have to unlock KeepassXC each time I open it. Some integration with Gnome Keyring would be nice.

**Comment by phoerious** [comment]
You cannot unlock KeePassXC with Gnome Keyring, but you can use it as a replacement for Gnome Keyring.

**Comment by danisztls** [comment]
What would be the rationality? Having Argon instead of SHA256? I suspect using this would lead to more integrations problems without solving any real issue.

**Comment by Aetf** [comment]
Putting aside whether using a PAM module to unlock KeepassXC is a good practice in terms of security or not, there are actually technical challenges implementing such a PAM module.

When I implemented the Fdo secret service provider in KeepassXC, automatically unlocking it using PAM was exactly the next step I planned to do next. But it turns out to be quite difficult if not possible to implement securely.

Fundamentally, the problem is that the PAM authentication process happens very early during user login. The user session doesn't even exist yet. So the PAM module has to find a way to store the password for a prolonged time until the user session is set up (with KeepassXC now running) and then unlock the database. This raises this question:

_How do you guarantee that you only give the password to KeepassXC, not something else happens to start quicker than KeepassXC and pretend to be KPXC?_

This rules out the possibility of directly using the kernel keyring:
* Using the user keyring basically makes the password readable by any program running as the current user
* Using the kernel root keyring with another service handle the unlocking faces the same challenge: how would the service verify the program it gives the password to is trusted to be KeepassXC.

The current implementation in Gnome keyring and KWallet work around this problem by using an IMHO ugly hack roughly like the following:

* login manager (SDDM, GDM, etc) starts the PAM stack, which calls into `pam_gnome.so` or `pam_kwallet.so`.
    + The PAM module saves the received password
    + It forks to another process and blocks wait on a pipe
* login manager sees that the PAM stack authenticates successfully and proceeds to create the user session
    + during the session set up, after environment variables are prepared (`DISPLAY`, `DBUS_SESSION_BUS_ADDRESS`, etc), a small service will send all current environment variables to the pipe
    + The forked process in the previous step receives and updates its environment variables and then `exec` into the actual keyring program with the password

The assumption is that we only trust a program we started ourselves. This trust model is also used in Wayland compositors, where certain privileged programs such as input methods are started by the compositor.

In this case, the keyring service is no longer a normal service managed by systemd or the DE's auto start mechanism. That's why I call it a hack and don't want to pursue it further. But if anyone else is interested, so far I feel replicating something like the above is the best we can do.

**Comment by danisztls** [comment]
@Aetf Thanks for the explanation.

**Comment by michaelk83** [comment]
Assuming such a PAM module is specific to KPXC, would it not be possible to set up some sort of encryption scheme (key pair? TOTP?) and/or communication protocol, so that only KPXC can read the password successfully? So that if another process attempts the same, it either fails to get anything, or only gets an encrypted password. Not because the other process is untrusted, but because it lacks some sort of shared secret.

This may also be related to [Linux fingerprint support](https://github.com/keepassxreboot/keepassxc/issues/5991) / [QuickUnlock](https://github.com/keepassxreboot/keepassxc/issues/488), because those features share a similar secure storage problem. Essentially, what we want to happen here is roughly analogous to the PAM module saving the password for QuickUnlock, and then when KPXC launches, it automatically would initiate the QuickUnlock. But the difference is that actual QuickUnlock has an extra security factor (user PIN / fingerprint), and here we don't (or at least not one that should require another user interaction).

And that also reminds me a related use case - what if a user wants to log in to the session using the fingerprint PAM module, and then use *that* to automatically unlock KPXC?

**Comment by Aetf** [comment]
> Assuming such a PAM module is specific to KPXC, would it not be possible to set up some sort of encryption scheme (key pair? TOTP?) and/or communication protocol, so that only KPXC can read the password successfully?

Not sure how this would work, given everything is readable by the user. So you can't embed any private secret in the program or the PAM module.

Automatic login or fingerprint unlock or anything that doesn't involve a password are not supported by Gnome keyring or KWallet. For KPXC I'm also not sure how to work around that.

**Comment by michaelk83** [comment]
> Not sure how this would work, given everything is readable by the user. So you can't embed any private secret in the program or the PAM module.

That's true. I was thinking that if it was somehow generated algorithmically, like a TOTP, then there won't be much to read, other than the binary code (or source code on github). It would have to be a purpose-built malware, and IIRC there was an argument that once the userspace is compromised with such malware, it's game over anyway.. But if the encryption is too weak, the password can be brute-forced without any specialized malware like that. A strong encryption would still require some sort of entropy source, which is most simply some sort of key file. Which brings us back to square one, so I guess that's not useful.

For the fingerprint SSO, the fingerprint becomes the password, I guess. So probably, the fingerprint reader PAM module needs to somehow pass that data to KPXC's PAM module, and then the latter would take over from there as usual.

But honestly, all of this is above my pay grade, so I'm just throwing ideas.

**Comment by droidmonkey** [comment]
Negative, a fingerprint is an identity (at best), aka username. You can use a fingerprint as a second factor to access a secure crypto store (similar to what Windows Hello and TouchID do), but you cannot use it as a password. Most fingerprint readers do this local to the reader, they then turn around and do FIDO2 with whatever is requesting it.

**Comment by michaelk83** [comment]
> a fingerprint is an identity (at best), aka username

Usually, yes. Yet somehow people use it to unlock their session, do they not? (If they choose to configure it that way.) Maybe that's more akin to a passwordless login with the _username_ replaced or verified by the fingerprint, but from the user's perspective the result is as same as if their fingerprint acted as a password. (edit: Or did you mean there is some sort of key file stored on the fingerprint reader, so the fingerprint is used to access that, and then the key file is used as the password? Even if so, that's still transparent to the user.)

Regardless, whatever data that comes from the fingerprint PAM will need to find its way to the KPXC PAM (once that exists), and be used there in whichever way it needs to be. Or something like that.

But before that, there's still the bigger issue that Aetf raised [above](https://github.com/keepassxreboot/keepassxc/issues/6055#issuecomment-1021682846).

**Comment by droidmonkey** [comment]
Closing this per Aetf's excellent feedback:

>When I implemented the Fdo secret service provider in KeepassXC, automatically unlocking it using PAM was exactly the next step I planned to do next. But it turns out to be quite difficult if not possible to implement securely.
>
>Fundamentally, the problem is that the PAM authentication process happens very early during user login. The user session doesn't even exist yet. So the PAM module has to find a way to store the password for a prolonged time until the user session is set up (with KeepassXC now running) and then unlock the database.

**Comment by michaelk83** [comment]
> That's why I call it a hack and don't want to pursue it further.

@Aetf , how about this, inspired by #6458 :
1. The PAM module would come with a client authorization file set to `root:root 600` (rw by root only). This prevents unauthorized clients from tampering with this file.
2. When KPXC is installed or upgraded, the system admin has to update this file (can be handled by the installer?), e.g.:
    `sudo sha256sum /absolute/path/to/keepassxc >> /etc/keepassxc/pam/clients`
    This produces a line in the form `<hash of binary> /absolute/path/to/keepassxc`.
3. During login, the PAM module saves the password e.g. in kernel memory, and waits.
4. KPXC is launched as normal (SystemD, Autostart, etc), and asks the PAM module for the password.
5. The PAM module detects the binary path and hash of the requesting process, similar to #6458 , and compares against the authorization file. If it finds a match, it returns the password, otherwise it refuses.
6. Optionally, the PAM module deletes the password from memory and shuts down after some timeout period.

The path and binary hash of KPXC is public knowledge, so it doesn't need to be hidden, but it's not something that another process can easily reproduce. As long as the authorization file is protected, I think this should work?

**Comment by droidmonkey** [comment]
There are a few people _asking_ for PAM integration and absolutely _zero_ people willing to code it. I certainly have no appetite to review the code, test it, and understand the myriad of potential security issues that can be had depending on each distribution. Just let this die please.

**Comment by Aetf** [comment]
What @michaelk83 said sounds like a plan. But I agree with @droidmonkey that we should close this.
People who want this can totally create a PAM module outside of KPXC repo. There's no need to host PAM code here I think.

**Comment by michaelk83** [comment]
> There's no need to host PAM code here I think.

Agreed.

There will still need to be the bit that polls the password from the PAM module, but that's minor. Depending on how QuickUnlock will eventually be implemented on Linux ( #5991 ), it might be possible to reuse (some of?) that code path, even. But yes, the main PAM module code doesn't need to be touched by KPXC team at all (and isn't even specific for KPXC).

**Comment by Patricol** [comment]
I doubt this is enough to re-open the issue; but my use case isn't subject to nearly any of the criticism leveled at this proposal.

I'm using KeePassXC exclusively as a replacement for KWallet/GnomeKeyring, specifically for ssh-agent, secretservice, and git (git-credential-keepassxc). I do not and will not use it to store any other passwords. I imagine users who do use KeePassXC as their primary password manager could just create a separate database to emulate my use case.

With its user-confirmation functionality for secretservice; KeePassXC is more secure than KWallet and GnomeKeyring.

PAM integration that securely achieves the goal of unlocking secretservice daemons using a login password may seem like a kludge because of the circumvention of systemd, but it's actually secure for users who cba to enter a separate (and thus likely less secure than login) password right after every login. Those users will insecurely use GnomeKeyring etc. to store the KeePassXC password that guards all of their primary passwords; creating some real risk. Much more real than if they just set their login password to match their KeePassXC password and used a 'kludgey' pam integration.

In summary, I believe even a kludgey pam integration would significantly improve security for users who attempt auto-login when considering that those users are going to settle for insecurity (often without realizing) rather than give up. I also present a use case where KeePassXC with a kludgey pam integration would be more secure than the standard usage patterns of the widely used KWallet/GnomeKeyring (which use the same kludgey pam integration).

**Comment by m00nwtchr** [comment]
Started working on something here: https://github.com/m00nwtchr/pam-keepassxc

**Comment by jm355** [comment]
Just chiming in to say that after a lot of times where gnome-keyring broke for me in sway, I switched to keepassxc and made an entirely separate db for my computer secrets, which I would like to unlock on login. I have it encrypted with a keyfile that only my user can read (`chmod u=r ~/path/to/key`), but it doesn't unlock automatically. When an app tries to access it, keepassxc pops up and I click unlock since it has the path to the keyfile stored, but it would be nice if I could set that db to attempt unlocking automatically. 

Is it 100% ideal security? No. But my hard drive is encrypted so it's safe when I power off, and I'm the only one who uses my computer so I'm not worried about other users or an admin getting the secrets, and if there's malware that can find that and do malicious things with the handful of desktop secrets I have, then it's likely already keylogged my normal database password and desktop secrets are the least of my worries

**Comment by m00nwtchr** [comment]
A more secure way to do that would probably be to use a TPM sealed secret, so that it cannot be stolen even while the disk is decrypted, but I think doing this with maximum possible security would require integration in KeePassXC itself rather than via an external program (like with some of my attempts)

**Comment by sumwale** [comment]
I implemented an entirely different, simpler approach to this: https://github.com/sumwale/pam-keepassxc. It doesn't depend on login password being related to the database password in any way, can unlock all the databases automatically on login or screen unlock. In fact user does not need to use password authentication at all and can go for completely passwordless login (fingerprint, SSO or what not). In other words it can be configured to never require password entry for any of the databases which is all handled automatically.

The approach is:

- Have the system maintain and provide the passwords for all the databases. Use a standard mechanism like `systemd-creds` for the encryption and management. This uses a host-specific key combined with TPM2 (if available).
- The PAM module will just start a system service specific to the user (e.g. for UID=1000 it will start `keepassxc-unlock@1000.service`) at the end of auth. This service will take care of unlocking all the registered databases for the user using KeePassXC's dbus interface. Then it will monitor the user's session on the system dbus to determine session/screen lock/unlock, active/inactive and session close. The service monitors the opened x11/wayland session and stops automatically when that session closes. So once started this service directly takes care of everything with no second module/process required in the middle.

Since there is no communication required with another process or with PAM for that matter, there are no hairy security issues around that. User-space malware cannot do anything since the service is running as root. Can't see any security issues with the approach, though I may be missing something. There are a few more details in the README on the github page. It is also easier for the users than quick unlock and I believe it is also more secure. If anything, users may be somewhat alarmed that no password has to be provided at all after the initial setup.

The code for the service is quite simple (https://github.com/sumwale/pam-keepassxc/blob/main/systemd/keepassxc-unlock). Right now this is just the first cut I was testing where the main service is a shell script which will be easy to convert to a C service (and a pam module, though it just launches the service so pam_exec seems enough for it). I have been running with this for a couple of days and is working well so far.

One issue is that if a user has multiple X/Wayland sessions running concurrently, then it can get all messed up. Don't know how dbus interface of KeePassXC will behave in such a case where multiple instances are running in different screens. In my current setup, switch user to the same user doesn't work but this used to work few releases back or in some other distro. Anyway this is something that will cause trouble right now and can be noted as unsupported.

@droidmonkey and others, will appreciate if you take a look and provide comments.

**Comment by m00nwtchr** [comment]
Nice, I wanted to implement storing credentials using TPM in my version but never got around to it. You can maybe link KPXC<-->User session by getting the session's PID from loginctl, and then checking for KPXC processes under that. But ideally only one instance of KPXC per user should be running since IIRC concurrent access to the database file isn't a good idea anyway.

**Comment by sumwale** [comment]
> You can maybe link KPXC<-->User session by getting the session's PID from loginctl, and then checking for KPXC processes under that.

Yes but how can the instance be specified using the KeePassXC dbus API? I guess only the first started instance will respond to the dbus API while others will not.

**Comment by m00nwtchr** [comment]
Depending on whether KeePassXC claims the DBus name "exclusively" they might get separate objects which can be selected from, but more likely only the first instance will actually be able to use the DBus api. Really it's just KeepPassXC not being designed for multiple sessions etc, though you could still detect this situation just to make sure nothing breaks. (only start 1 instance of keepassxc and so on)

**Comment by droidmonkey** [comment]
We advertise on the session bus, not the system bus, for our own DBUS API. So each user should have their own capability to communicate independently to their instance of keepassxc. I am not a dbus expert, but that's my understanding.

**Comment by sumwale** [comment]
> So each user should have their own capability to communicate independently to their instance of keepassxc.

Right, the question is what happens if a user is running multiple instances of keepassxc. How does the D-Bus API work in that case?

Edit: just tested it, and as expected only the first instance responds to the dbus calls

**Comment by sumwale** [comment]
> though you could still detect this situation just to make sure nothing breaks. (only start 1 instance of keepassxc and so on)

Not sure how. Relying on the executable name? Doesn't seem reliable. Does KeePassXC itself use some lock file or similar to detect existing instance (for the "Start only a single instance ..." option).

**Comment by michaelk83** [comment]
> Right, the question is what happens if a user is running multiple instances of keepassxc. How does the D-Bus API work in that case?

If the instances are running as the *same* user, then the first one to run is the one that grabs the bus address. IIRC, KeePassXC has code to prevent multiple instances for the same user (as it doesn't make a whole lot of sense to run multiple instances like that), so that's not something you need to worry about. You can just write it off as a limitation of your system.

If the instances are running as *different* users, each user should have their own session bus, so each instance will have an address on the corresponding separate bus.

See https://stackoverflow.com/a/71254213/15047974

**Comment by sumwale** [comment]
> If the instances are running as different users, each user should have their own session bus

Right obviously. The question was about multiple instances started by the same user (see the last para in my original note).

If you get a chance, please do take a look at the original [comment](https://github.com/keepassxreboot/keepassxc/issues/6055#issuecomment-2746529473) and code to see if I have overlooked anything major.

**Comment by michaelk83** [comment]
> If you get a chance, please do take a look at the original [comment](https://github.com/keepassxreboot/keepassxc/issues/6055#issuecomment-2746529473) and code

I did read the comment, sounds very interesting. A bit too busy to check the code atm, but will try to get to it later. From your last paragraph I did understand it as multiple *user* sessions (so separate buses). Are you saying that Wayland may start multiple sessions for the same user?

> switch user to the same user
> in different screens

Hmm... Didn't know that was possible.

Still, why would you run multiple KeePassXC like that?

**Comment by sumwale** [comment]
> Are you saying that Wayland may start multiple sessions for the same user?

Don't know about Wayland but X11 certainly does allow it. Anyway it is a very niche usage but just noted it for completeness.

**Comment by droidmonkey** [comment]
If the user is running multiple instances of keepassxc, then they will have to deal with the "consequences" of that. We enable single instance by default for good reasons.

**Comment by m00nwtchr** [comment]
On systemd systems, multiple sessions of the same user share the same dbus bus. I don't know exactly what the mechanism the single instance toggle in KPXC uses though.

@michaelk83 
"Starting multiple sessions" is just going to a different TTY, logging in, and starting a X server/Wayland compositor there, nothing more.

**Comment by sumwale** [comment]
I finally got sufficient time to finish the C version of the code: https://github.com/sumwale/keepassxc-unlock. The current implementation does not use PAM at all, rather just a background system service that watches session logins and launches an unlocker service for new X11/Wayland sessions. It also has checksum verification of keepassxc executable for some additional confirmation before unlocking. In theory this approach will also work on Windows/Mac if they have something equivalent to the D-Bus API that allows for KeePassXC unlock by a system service.

@droidmonkey  and others: will you guys accept a PR that integrates this into the main KeePassXC codebase? Specifically the integration will take care of the unlock setup from within the UI:

* During database creation and in settings, a button/checkbox can enable/disable automatic unlock of the database. This will use polkit to ask the system service to write the configuration for unlock. This is currently done explicitly by running keepassxc-unlock-setup, but integration into the main UI will provide a much better experience to the end users.
* At startup, keepassxc can check the last recorded timestamp and hash in its cache, then ask user to re-register if it has changed (and the database is marked for auto-unlock). This is for the checksum verification of the binary which will fail if it changes due to app upgrade. Btw, though there is checksum verification, I don't believe it provides any additional security, so the whole thing can be removed after discussion. If an app has enough privileges to displace keepassxc on D-Bus, then it can simply launch a patched keepassxc and easily fool the user into giving the password in any case.

If you guys will accept such a PR, then I can start looking into KeePassXC code and raise a PR in a few days.

The daemon code can be moved to the KeePassXC codebase or live outside as you guys prefer. Native packaging (and I guess snap with classic mode) can create a separate package for the daemons that the main package can recommend. Others can install the static binaries as I have on that github page.

**Comment by droidmonkey** [comment]
Honestly we will not want to incorporate this, there is no way I can maintain this code and ensure it works. You are welcome to offer it as a fork, but incorporating this into the official release is not in our best interest and also conflicts with our desire to not "auto unlock" things. One of the main reasons people use KeePassXC over the built-in password manager or gnome keyring is a separation between your computer login and your secrets login.

**Comment by sumwale** [comment]
@droidmonkey have you read my comments before? In this implementation there is no relation between the login password and the keepassxc passwords. In fact the login may be using a completely different mechanism having no password like fingerprint, SSO or whatever. This can also auto-unlock any number of databases and not just one. It is also more secure than current quick unlock in my opinion.

> Have the system maintain and provide the passwords for all the databases. Use a standard mechanism like systemd-creds for the encryption and management. This uses a host-specific key combined with TPM2 (if available).

Please see the comment before https://github.com/keepassxreboot/keepassxc/issues/6055#issuecomment-2746529473

The current implementation has made the change that it has also gotten rid of PAM module completely, and rather has a global service that watches session logins on the system D-Bus. So users don't have to muck around with the PAM configurations anymore. Just install, register databases that need to be unlocked and that's it. If this was integrated in the UI then it will be much easier (in that registration can happen right from the UI).

**Comment by droidmonkey** [comment]
You are welcome to incorporate the code and draft a PR. I'll take a look at it when possible, but I am wary of "yet another Linux feature" that becomes super complicated across the 15 distros and sandboxes.

**Comment by michaelk83** [comment]
IMHO, this should be kept as a separate tool. It goes against the security best practice of keeping the DB locked when it's not in immediate use (for which purpose there are the various auto-lock features in KeePassXC), and unlocking only when needed.

The default setup recommendation should be a strong passphrase + auto-lock + QuickUnlock to unlock when needed. If a user wants to use the unlock tool for their own convenience, it's on them to install and manage it (same as any other non-standard setup).

> * At startup, keepassxc can check the last recorded timestamp and hash in its cache, then ask user to re-register if it has changed (and the database is marked for auto-unlock).

Since KeePassXC runs as the normal user, and the hash needs to be saved as root, this isn't very helpful except as a reminder. It just adds one more prompt that the user may habitually dismiss without even looking at it properly.

> * During database creation and in settings, a button/checkbox can enable/disable automatic unlock of the database.

This is also of limited use, IMO, since it only needs to be done once or twice ever. There's already some separate technical action required to install the unlock tool. Having the user run `keepassxc-unlock-setup` doesn't add much complexity. If `keepassxc-unlock-setup` is difficult to use, you can instead have it guide the user when it's run with default options.

Moreover, this UI setting would need to be per-DB, but DBs are portable across devices. This would present the difficulty: how to handle this setting when the DB is copied to another device? It would be simpler to just have the user run `keepassxc-unlock-setup` where needed.

Regarding security concerns:
- To displace KeePassXC on D-Bus, all that's required is just run earlier than KeePassXC. There are no special privileges required.
- If you have security concerns re QuickUnlock, you can post an issue to discuss those concerns and/or submit a PR to address them.

**Comment by sumwale** [comment]
> It goes against the security best practice of keeping the DB locked when it's not in immediate use

How so? It unlocks the DB only on initial start of keepassxc and on screen unlock. Are you saying that users should unlock keepassxc only before use? That would be a weird way of using a password manager of having to unlock it first (using password in majority of cases) before reading from it every time. It will also not work for the secret service impl which will silently fail if the db is locked.

> Since KeePassXC runs as the normal user, and the hash needs to be saved as root, this isn't very helpful except as a reminder. 

No, this will work the same way the setup in UI integration would work. It will use polkit for user auth and use the D-Bus service provided by the unlock service to update the hash.

> This is also of limited use, IMO, since it only needs to be done once or twice ever.

It provides a better/easier guarantee that it is the correct keepassxc process for the purpose of recording the hash. The other reason for inclusion would be that native packages can build the unlock service as a package from the keepassxc sources rather than from an unknown third party.

> Moreover, this UI setting would need to be per-DB, but DBs are portable across devices.

That does present a problem. I guess it can be recorded only in .cache rather that the database.

> To displace KeePassXC on D-Bus, all that's required is just run earlier than KeePassXC. There are no special privileges required.

That is why the hash check was added which will work in a more secure way from inside the GUI where the app that registers its hash also provides the password to be recorded.

In any case if an app/malware has gained enough privileges to displace keepassxc on D-Bus then all bets are off. For example such an app can displace keepassxc itself and start its own patched version of keepassxc that sends/records all the passwords elsewhere.  The hash check helps with only a small set of cases e.g. malware in sandbox with limited privileges. The only way to completely avoid such cases would be to use MAC+IMA+EVM on Linux and equivalent on other platforms, which is obviously beyond the scope of keepassxc.

**Comment by michaelk83** [comment]
> Are you saying that users should unlock keepassxc only before use? That would be a weird way of using a password manager of having to unlock it first (using password in majority of cases) before reading from it every time.

It depends on how often one needs to access passwords from the manager, but yes, typically I would imagine cycles of unlock &rarr; access the passwords you need &rarr; lock. You don't leave your safe open all day just in case you might need to take some cash or jewelry out at some later point. Unless you're constantly swapping stuff in and out. And the same way, you don't work in `sudo` all day. You escalate privileges only when needed, and de-escalate as soon as you don't need them anymore. This is what QuickUnlock is for.

Personally, I use SSO for my daily (low-security) web accounts, so I need to access passwords very rarely. Though users like myself can simply not enable auto-unlock. I can't attest to other users' use cases or patterns.

> It will also not work for the secret service impl which will silently fail if the db is locked.

Secret Service has a mechanism to request DB unlock when a password is needed. Sadly, there are some issues with it due to oversights in the API design, so it doesn't always work as intended. But even with the auto-unlock it can still be racy. If KeePassXC is auto-launched by DBus when the Secret Service endpoint is first accessed, then it's a race between the native unlock prompt, your auto-unlock, and whatever timeouts the client app may experience if it's hasn't implemented the unlock sequence correctly.

> In any case if an app/malware has gained enough privileges to displace keepassxc on D-Bus then all bets are off.

Like I said, there are no special priviliges needed for that. You can install a user-space malware and edit user-space auto-start config to have it run before KeePassXC, and grab the DBus address. You don't need root or anything.

Which is to say, yes, the hash check can be helpful, somewhat. But generally speaking, I agree, if you get malware on your PC, all bets are off.

**Comment by sumwale** [comment]
> You escalate privileges only when needed, and de-escalate as soon as you don't need them anymore. This is what QuickUnlock is for.

Is this a recommendation in the keepassxc docs or just your personal opinion? It makes no sense at all. If a malicious app can read the process memory to obtain the database password, it can do so after the db has been unlocked. This pattern of usage provides no security benefit albeit does significantly reduce the usability. It is a password manager not a vault. Besides quick unlock is not useful for a large proportion of users who are using just password based login.

> If KeePassXC is auto-launched by DBus when the Secret Service endpoint is first accessed

It will almost always be auto-launched on login like the option provided in keepassxc settings.

> You can install a user-space malware and edit user-space auto-start config to have it run before KeePassXC, and grab the DBus address. You don't need root or anything.

As I mentioned: 1) malware can just launch a patched version of keepassxc that grabs passwords so that the user has no way to differentiate it from the genuine one and will just type in the db password, 2) the hash check avoids the above in any case. This has nothing to do with root. There is no additional risk posed due to the D-Bus call.

**Comment by michaelk83** [comment]
> Is this a recommendation in the keepassxc docs or just your personal opinion? ...

Primarily my own opinion, backed by various sources I've read over the years; plus my interpretation of KeePassXC's intention, having followed the project for a long time now. I can't speak re the dev's official stance, but whether they agree with me or not, it's not something I'd expect to find in the docs.

Fundamentally, when the DB is unlocked, there are multiple attack vectors potentially possible. KeePassXC tries its best to protect against most of them, but nothing is 100% hermetic. When the DB is locked, the attack surface is significantly reduced. And likewise, the less time it's kept unlocked, the smaller the cumulative attack surface.

In the end, as always, it's a choice between security and convenience. It's up to each user to decide which balance point fits their use case. If this pattern doesn't make sense to you, then go ahead and work in `sudo` all day. Or choose whatever other balance point that you think makes sense for you.

> quick unlock is not useful for a large proportion of users who are using just password based login.

You don't need anything else for QuickUnlock. You should be able to set it up to use a PIN if you don't have biometrics or whatnot. So you enter your passphrase on the 1st unlock, and use whatever you've set up QuickUnlock with on subsequent unlocks. It's a middle ground between keeping your DB unlocked all the time, and using your full passphrase (etc) all the time.

> It will almost always be auto-launched on login like the option provided in keepassxc settings.

I don't see any basis for that claim. This is very much user-dependent. If KeePassXC is used mostly for Secret Service, having it launch on first access via DBus is a common setup. (Technically, that would still often happen close to login, as various client apps tend to be set up for auto-start. But it doesn't have to be.)

If KeePassXC is used for other things, then it's up to each user's use case when they need it to be available.

> As I mentioned ...

This part is going in circles. We are basically in agreement, just looking at it from different angles.

**Comment by sumwale** [comment]
> When the DB is locked, the attack surface is significantly reduced. And likewise, the less time it's kept unlocked, the smaller the cumulative attack surface.

A proper security analysis will take the precise parameters into account rather than vague assertions about the attack surface. In this case there is exactly one variable, namely the plain-text password in process memory. This variable remains the same in either of the usage patterns. Keeping the database locked for longer duration only reduces the duration of attack window but does not help block any of the attack vectors, and thus in effect the attack surface remains the same.

The only case that can be argued is if a user frequently fails to lock the screen when leaving it unattended. It will be a real stretch to imagine users who habitually lock the keepassxc databases but do not lock the screen.

> You should be able to set it up to use a PIN if you don't have biometrics or whatnot.

Oh, is this a feature in the quick unlock implementation, or you mean using an appropriate PAM configuration? Anyway quick unlock is neither as secure as this auto-unlock implementation (because it stores the key in process+kernel memory and thus is vulnerable to attacks like Thunderspy), nor is as convenient.

**Comment by michaelk83** [comment]
> is this a feature in the quick unlock implementation, or you mean using an appropriate PAM configuration?

QuickUnlock doesn't use PAM, IIRC. It does use Polkit to request the QuickUnlock credential. You should be able to set it up in the system settings to request a PIN. There was also a request to use a PIN directly from KeePassXC, but it's not merged yet, I think.

You can look at the implementation for further details, as I don't remember them all that well. And like I said earlier, if you have security or other concerns re QuickUnlock, you're welcome to submit an issue of PR to address them. E.g a TPM integration would likely be welcome.

**Comment by jm355** [comment]
> Sorry if I missed something, but what do you mean by "Quick unlock"?

Looks like they're maybe talking about this? https://keepassxc.org/docs/KeePassXC_UserGuide#_quick_unlock but those instructions seem specific to Mac OS and Windows, while this discussion is related to linux, so I don't understand why it's being talked about as fulfilling the request of this ticket.

**Comment by michaelk83** [comment]
> I don't understand why it's being talked about as fulfilling the request of this ticket.

QuickUnlock isn't meant to fulfill this ticket directly, but it is an alternative. The Linux implementation is in #8983, which I now see is not released yet. It's available in the snapshot builds.

Specifically for this ticket, you can use [sumwale's auto-unlock tool](https://github.com/sumwale/keepassxc-unlock). The recent discussion was on whether or not it should be integrated more intimately into KeePassXC (and also talked about QuickUnlock as an alternative).

**Comment by variabl3** [comment]
I'd just like to say that this is one of the most-interesting and informative threads I've read in a while.  Thank you!

**Comment by benlavender** [comment]
Yes this is a good read, I've been wanting auto login of KPXC for use with the secret service for simple tokens and random passwords generated by applications and a separate (manual login) for my own secrets. I know I can use `AutoOpen` to unlock that first database for use via secrets api from manual unlock of the second one. Problem I get is having to manually unlock the databases on login and I get quirks where prompts come up to create a new database each time. 

Perhaps I'm best just using kwallet with KDE for tokens and keep KPXC for my own secrets 🤷‍♂️
